Fix Issue #23801

ejwessel opened the pull request:
<!--
Please submit this PR as a draft initially.
Do not mark it as "Ready for review" until the template has been completely filled out, and PR status checks have passed at least once.
-->

## **Description**

#23801

<!--
Write a short description of the changes included in this pull request, also include relevant motivation and context. Have in mind the following questions:
1. What is the reason for the change?
2. What is the improvement/solution?
-->

[![Open in GitHub Codespaces](https://github.com/codespaces/badge.svg)](https://codespaces.new/MetaMask/metamask-extension/pull/24150?quickstart=1)

## **Related issues**

Fixes:
<img width="1800" alt="Screenshot 2024-04-20 at 3 07 07 PM" src="https://github.com/MetaMask/metamask-extension/assets/3590816/9e7bc251-add6-4ba3-a7cb-1c83498ace86">

## **Manual testing steps**

1. Go to this page...
2.
3.

## **Screenshots/Recordings**

<!-- If applicable, add screenshots and/or recordings to visualize the before and after of your change. -->

### **Before**

<!-- [screenshots/recordings] -->

### **After**

<!-- [screenshots/recordings] -->

## **Pre-merge author checklist**

- [ ] I’ve followed [MetaMask Coding Standards](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/CODING_GUIDELINES.md).
- [ ] I've completed the PR template to the best of my ability
- [ ] I’ve included tests if applicable
- [ ] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [ ] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

darkwing committed:
UX: Multichain: Release the send page

Restore domain-input

Fix test for send error

Fix max button test

E2Es: Update amount selector, Next to Continue button

Remove header for send flow

Fix address book and ENS tests

Update snap-account-transfers.spec.ts and helpers.js

Update change-language.spec.js

Fix custom-token-send-transfer.spec.js

Fix missing NFT icon

Make send-nft.spec.js pass, set send volume to 1 when NFT

Fix send-edit.spec.js

Fix address-book.spec.js

Release the address book changes

darkwing committed:
Restore localizations

darkwing committed:
Fixes #22986 - UX - Multichain - Send flow - Show address in account … (#23025)

The shortened address should display in the AccountPicker within the new
Send flow

Fixes: https://github.com/MetaMask/metamask-extension/issues/22986

1. Open the new send flow
2. See the account address under the account name

<!-- If applicable, add screenshots and/or recordings to visualize the
before and after of your change. -->

N/A

<img width="420" alt="SCR-20240216-mxwm"
src="https://github.com/MetaMask/metamask-extension/assets/46655/ca6dbe2b-c136-474a-860d-970ccdad9f86">

- [ ] I’ve followed [MetaMask Coding
Standards](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/CODING_GUIDELINES.md).
- [ ] I've clearly explained what problem this PR is solving and how it
is solved.
- [ ] I've linked related issues
- [ ] I've included manual testing steps
- [ ] I've included screenshots/recordings if applicable
- [ ] I’ve included tests if applicable
- [ ] I’ve documented my code using [JSDoc](https://jsdoc.app/) format
if applicable
- [ ] I’ve applied the right labels on the PR (see [labeling
guidelines](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/LABELING_GUIDELINES.md)).
Not required for external contributors.
- [ ] I’ve properly set the pull request status:
  - [ ] In case it's not yet "ready for review", I've set it to "draft".
- [ ] In case it's "ready for review", I've changed it from "draft" to
"non-draft".

- [ ] I've manually tested the PR (e.g. pull and build branch, run the
app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described
in the ticket it closes and includes the necessary testing evidence such
as recordings and or screenshots.

darkwing committed:
Fix #22938 - UX - Multichain - Send Flow - Update ENS address upon network change (#23024)

## **Description**

Changing the network mid-send can effect ENS domains in a negative way.
Namely, an ENS domain can have a different addresses per chain. This PR
ensure that ENS lookups are done when the user changes chains.

## **Related issues**

Fixes:  https://github.com/MetaMask/metamask-extension/issues/22938

## **Manual testing steps**

### Scenario 1
0.  Have an ENS with two different address for Mainnet and Goerli
1.  Enter the new send flow
2. Type in the ENS on Mainnet -- see a search result with the correct
address
3. Change the network picker to Goerli
4. See the address change in the search result

### Scenario 2
0.  Have an ENS with two different address for Mainnet and Goerli
1.  Enter the new send flow
2. Type in the ENS on Mainnet -- see a search result with the correct
address
3. Click the search result (you'll go to the Amount screen)
4. Change the network picker to Goerli
5. See the address change in the recipient field change
6. Execute the transaction
7. Ensure token sent to correct address

## **Screenshots/Recordings**

<!-- If applicable, add screenshots and/or recordings to visualize the
before and after of your change. -->

### **Before**

(In the GitHub issue)

### **After**

Coming....

## **Pre-merge author checklist**

- [ ] I’ve followed [MetaMask Coding
Standards](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/CODING_GUIDELINES.md).
- [ ] I've clearly explained what problem this PR is solving and how it
is solved.
- [ ] I've linked related issues
- [ ] I've included manual testing steps
- [ ] I've included screenshots/recordings if applicable
- [ ] I’ve included tests if applicable
- [ ] I’ve documented my code using [JSDoc](https://jsdoc.app/) format
if applicable
- [ ] I’ve applied the right labels on the PR (see [labeling
guidelines](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/LABELING_GUIDELINES.md)).
Not required for external contributors.
- [ ] I’ve properly set the pull request status:
  - [ ] In case it's not yet "ready for review", I've set it to "draft".
- [ ] In case it's "ready for review", I've changed it from "draft" to
"non-draft".

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the
app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described
in the ticket it closes and includes the necessary testing evidence such
as recordings and or screenshots.

darkwing committed:
Fix test/e2e/tests/send-eth.spec.js

darkwing committed:
Fix send-nft.sepc.js

darkwing committed:
Fix lint in routes test

darkwing committed:
Rebased test helper fix

darkwing committed:
Fix send-hex-address.spec.js

darkwing committed:
Fix custom-token-send-transfer.spec.js

darkwing committed:
Fix #23304 - Only use Confusable during send with ENS

darkwing committed:
Fix toChecksumHexAddress import

darkwing committed:
Fix gas tests

darkwing committed:
Remove more feature flags from test

darkwing committed:
Fix lint

darkwing committed:
Add back data-id for asset picker error

Commit:
docs: add merge train plan

Commit:
docs: update plan

darkwing committed:
Revert "Fix #22938 - UX - Multichain - Send Flow - Update ENS address upon network change (#23024)"

This reverts commit f17a26b9810150f6bccb665e6eb4774a8eda4a3e.

darkwing committed:
Remove unwanted, incomplete test

Commit:
chore: create util for swap+send quotes

Commit:
fix: clear quote data when switching to send

Commit:
docs: add task

Commit:
fix: update deps

requires useCallback for drilled prop
requires non-curried functions

Commit:
docs: add angela's suggestions

darkwing committed:
Restore @micaelae's useTokenTracker fix

Commit:
chore: mute borders

Commit:
chore: make icon smaller

Commit:
chore: align form items

darkwing committed:
Remove confirmation of sendScreenToConfirmScreen to avoid PPOM issues

darkwing committed:
Revert localization changes

Commit:
fix: disable swaps on unsupported networks

Commit:
chore: update sender account name styling

Commit:
chore: reduce converted text size

simplify styling

Commit:
feat: show recipient address

Commit:
chore: center div

Commit:
chore: remove unneeded padding

Commit:
chore: make tab text reponsive

Commit:
chore: update modal font weights

Commit:
chore: maximum token list real estate

move search bar
take up full height

Commit:
chore: remove unneeded dep

Commit:
chore: prevent scroll wiggle

update flexbox stylings

darkwing committed:
UX: Multichain: Send Flow: Disable account picker for edits (#23708)

## **Description**

Disables the account picker in the new send flow during an Edit

[![Open in GitHub
Codespaces](https://github.com/codespaces/badge.svg)](https://codespaces.new/MetaMask/metamask-extension/pull/23708?quickstart=1)

## **Related issues**

Fixes: N/A

## **Manual testing steps**

1. `yarn start`
2. Start a send
3. Continue
4. Click Edit to edit the send
5. See the account picker disabled

## **Screenshots/Recordings**

### **Before**

N/A

### **After**

<img width="573" alt="SCR-20240325-mnda"
src="https://github.com/MetaMask/metamask-extension/assets/46655/ad185a0c-8fc3-4174-95de-61731f1163e6">


## **Pre-merge author checklist**

- [ ] I’ve followed [MetaMask Coding
Standards](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/CODING_GUIDELINES.md).
- [ ] I've completed the PR template to the best of my ability
- [ ] I’ve included tests if applicable
- [ ] I’ve documented my code using [JSDoc](https://jsdoc.app/) format
if applicable
- [ ] I’ve applied the right labels on the PR (see [labeling
guidelines](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/LABELING_GUIDELINES.md)).
Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the
app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described
in the ticket it closes and includes the necessary testing evidence such
as recordings and or screenshots.

Commit:
chore: update deps

BZahory committed:
feat: create layout for swap and send (#23741)

<!--
Please submit this PR as a draft initially.
Do not mark it as "Ready for review" until the template has been
completely filled out, and PR status checks have passed at least once.
-->

## **Description**

<!--
Write a short description of the changes included in this pull request,
also include relevant motivation and context. Have in mind the following
questions:
1. What is the reason for the change?
2. What is the improvement/solution?
-->

[![Open in GitHub
Codespaces](https://github.com/codespaces/badge.svg)](https://codespaces.new/MetaMask/metamask-extension/pull/23741?quickstart=1)

## **Related issues**

Fixes:

## **Manual testing steps**

1. Go to this page...
2.
3.

## **Screenshots/Recordings**

<!-- If applicable, add screenshots and/or recordings to visualize the
before and after of your change. -->

### **Before**

<!-- [screenshots/recordings] -->

### **After**

<!-- [screenshots/recordings] -->

## **Pre-merge author checklist**

- [ ] I’ve followed [MetaMask Coding
Standards](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/CODING_GUIDELINES.md).
- [ ] I've completed the PR template to the best of my ability
- [ ] I’ve included tests if applicable
- [ ] I’ve documented my code using [JSDoc](https://jsdoc.app/) format
if applicable
- [ ] I’ve applied the right labels on the PR (see [labeling
guidelines](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/LABELING_GUIDELINES.md)).
Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the
app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described
in the ticket it closes and includes the necessary testing evidence such
as recordings and or screenshots.

---------

Co-authored-by: David Walsh <davidwalsh83@gmail.com>

Commit:
feat: show all swap tokens

Commit:
chore: only call API if properties are missing

sahar-fehri commented:
I haven't tested yet but im wondering if similar issue can happen for ERC721?

BZahory committed:
feat: integrate swap+send data flows (#23835)

<!--
Please submit this PR as a draft initially.
Do not mark it as "Ready for review" until the template has been
completely filled out, and PR status checks have passed at least once.
-->

<!--
Write a short description of the changes included in this pull request,
also include relevant motivation and context. Have in mind the following
questions:
1. What is the reason for the change?
2. What is the improvement/solution?
-->

[![Open in GitHub
Codespaces](https://github.com/codespaces/badge.svg)](https://codespaces.new/MetaMask/metamask-extension/pull/23835?quickstart=1)

Fixes:

1. Go to this page...
2.
3.

<!-- If applicable, add screenshots and/or recordings to visualize the
before and after of your change. -->

<!-- [screenshots/recordings] -->

<!-- [screenshots/recordings] -->

- [ ] I’ve followed [MetaMask Coding
Standards](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/CODING_GUIDELINES.md).
- [ ] I've completed the PR template to the best of my ability
- [ ] I’ve included tests if applicable
- [ ] I’ve documented my code using [JSDoc](https://jsdoc.app/) format
if applicable
- [ ] I’ve applied the right labels on the PR (see [labeling
guidelines](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/LABELING_GUIDELINES.md)).
Not required for external contributors.

- [ ] I've manually tested the PR (e.g. pull and build branch, run the
app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described
in the ticket it closes and includes the necessary testing evidence such
as recordings and or screenshots.

---------

Co-authored-by: David Walsh <davidwalsh83@gmail.com>
Co-authored-by: micaelae <100321200+micaelae@users.noreply.github.com>

salimtb commented on ui/ducks/send/send.js:
```diff
@@ -1420,6 +1420,13 @@ const slice = createSlice({
+          draftTransaction.sendAsset.details.standard === TokenStandard.ERC1155 &&
+          draftTransaction.amount.value === '0x0':
+            draftTransaction.amount.error = INSUFFICIENT_TOKENS_ERROR
+          break;
```
can you add a unit test for this case ?

darkwing committed:
Fix jest

sleepytanya commented:
@sahar-fehri  I haven't seen this issue with ERC721 yet. Good reminder! Will double-check.

sleepytanya commented:
@sahar-fehri I can't send ERC721 - every attempt results in frozen MM (429 errors) before I can actually see the next screen.

https://github.com/MetaMask/metamask-extension/assets/104780023/50b2966e-e3d2-4572-9584-66c5144a021e


The fix is working great!


https://github.com/MetaMask/metamask-extension/assets/104780023/1b4d933c-d387-4394-9fcd-e515bb2f908c


May be it's worth asking @hesterbruikman what message we want to display: 'Insufficient tokens' or 'Cannot send negative or zero amount of tokens'.

Commit:
chore: merge files after rebase

Commit:
fix: prevent flicker

pull out asset list logic
split up into to and from token lists
limit based on number of unowned tokens

Commit:
refactor: extract hook

Commit:
feat: add swaps refresh cadence to send

Commit:
fix: normalize API value

Commit:
feat: fetch uncached token spot prices

Commit:
fix: update submit disable predicate

Commit:
fix: remove input padding

Commit:
docs: update task list

BZahory committed:
feat: finish swap+send UI and handle transactions (#23889)

<!--
Please submit this PR as a draft initially.
Do not mark it as "Ready for review" until the template has been
completely filled out, and PR status checks have passed at least once.
-->

<!--
Write a short description of the changes included in this pull request,
also include relevant motivation and context. Have in mind the following
questions:
1. What is the reason for the change?
2. What is the improvement/solution?
-->

[![Open in GitHub
Codespaces](https://github.com/codespaces/badge.svg)](https://codespaces.new/MetaMask/metamask-extension/pull/23889?quickstart=1)

Fixes:

1. Go to this page...
2.
3.

<!-- If applicable, add screenshots and/or recordings to visualize the
before and after of your change. -->

<!-- [screenshots/recordings] -->

<!-- [screenshots/recordings] -->

- [ ] I’ve followed [MetaMask Coding
Standards](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/CODING_GUIDELINES.md).
- [ ] I've completed the PR template to the best of my ability
- [ ] I’ve included tests if applicable
- [ ] I’ve documented my code using [JSDoc](https://jsdoc.app/) format
if applicable
- [ ] I’ve applied the right labels on the PR (see [labeling
guidelines](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/LABELING_GUIDELINES.md)).
Not required for external contributors.

- [ ] I've manually tested the PR (e.g. pull and build branch, run the
app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described
in the ticket it closes and includes the necessary testing evidence such
as recordings and or screenshots.

---------

Co-authored-by: micaelae <100321200+micaelae@users.noreply.github.com>

sahar-fehri commented:
Thanks @sleepytanya ! The 429 you got is curious, i wasnt able to reproduce, and it seems okay for ERC721, the UI will display "balance: 1 token"

https://github.com/MetaMask/metamask-extension/assets/10994169/6b48e991-d475-4957-b205-cc2cafb313f5

legobeat commented:
@sahar-fehri 429 Rate-Limiting errors are common for users coming through e.g. Tor or commercial VPN services. As an aside, in general we shouldn't assume availability of these endpoints and fail gracefully when they aren't available.

darkwing committed:
remove flag check from test/e2e/tests/ppom/ppom-blockaid-alert-simple-send.spec.js

Commit:
fix: swap token lists

Commit:
fix: only reset gas limit if no swap

Commit:
chore: use decimal param for chain id

Commit:
chore: use swaps dev api

Commit:
fix: update function parameters

Commit:
fix: fetch single token balance when selected

Commit:
fix: gracefully handle errors for balance

Commit:
chore: reduce balance api calls

Commit:
refactor: remove api-calling component

Commit:
fix: use native currency for conversion

Commit:
chore: improve token caching

Commit:
fix: prevent crash

Commit:
fix: check for non-zero decimal

Commit:
chore: hide hex data field for swap+send

Commit:
chore: add basic func dep (resolves MetaMask-planning#2349)

Commit:
chore: add quote validation

Commit:
refactor: reorder switch cases

Commit:
refactor: change variable type

Commit:
chore: add loading animation

Commit:
chore: add undefined guard

Commit:
feat: handle approvals

i lost this in a rebase i think

Commit:
chore: update text formatting

Commit:
feat: update quote card styling

add TOS
add fee text
add scroll anchor

Commit:
fix: disable submit when loading

ejwessel commented:
Unrelated to my change, the tests are inconsistent and often fail when an address is selected or inserted as the recipient.

Common errors are:
```
-----Received an error from Chrome-----
SES_UNHANDLED_REJECTION: (TypeError#2)
----------End of Chrome error----------

-----Received an error from Chrome-----
TypeError#2: Cannot set properties of undefined (setting 'quotes')
----------End of Chrome error----------
```
```
-----Received an error from Chrome-----
(BigNumber Error#1)
----------End of Chrome error----------

-----Received an error from Chrome-----
BigNumber Error#1: new BigNumber() not a number: 2540be400
----------End of Chrome error----------
```

Commit:
refactor: pull out ref

Commit:
chore: prevent underline

Commit:
fix: update time formatting

Commit:
chore: update alignment

Commit:
feat: add `disableUnderline` to `ButtonLink`

Commit:
chore: remove underline on link

Commit:
fix: reduce fetches

ejwessel committed:
chore: use waitForSelector

Commit:
docs: add task

BZahory committed:
fix: prevent user from sending 0 NFTs (#24420)

Signed-off-by: Ethan Wessel <ethan.wessel@consensys.net>
Co-authored-by: Ethan Wessel <ethan.wessel@consensys.net>

Commit:
fix: show fiat 0 when balance 0

Commit:
chore: prevent bounce in styling

Commit:
feat: auto-submit approval tx

Commit:
fix: remove unneeded padding

ejwessel commented:
Change was cherry picked out.

georgewrmarshall commented on ui/components/component-library/button-icon/button-icon.tsx:
```diff
@@ -20,6 +20,7 @@ import {
 } from './button-icon.types';
 
 const buttonIconSizeToIconSize: Record<ButtonIconSize, IconSize> = {
+  [ButtonIconSize.Xs]: IconSize.Xs,
```
Hey @ejwessel, this method for adding a size seems correct, but changes to the component-library require a thorough review to ensure alignment with the Figma and mobile component. It also appears that this update isn't currently being used. Should this be removed?

Notice the Figma component does not have an Xs size

![Screenshot 2024-05-23 at 8 32 06 AM](https://github.com/MetaMask/metamask-extension/assets/8112138/43b36261-f451-46a1-a583-1b4061f736b5)

georgewrmarshall commented on ui/components/multichain/pages/send/components/quote-card/index.tsx:
```diff
@@ -0,0 +1,236 @@
+        href={CONSENSYS_TERMS_OF_USE}
+        target="_blank"
+        className="quote-card__TOS"
+        disableUnderline
```
If the underline on hover must be overridden can we do it with a custom className instead of updating the component api itself?

georgewrmarshall reviewed (commented):
Hey @ejwessel, I've started a discussion in slack regarding component library changes https://consensys.slack.com/archives/C0354T27M5M/p1716478907218739
